File: components/user/user-service.ts
import {
  createUserWithEmailAndPassword,
  deleteUser,
  updateProfile,
} from 'firebase/auth';
import { doc, setDoc } from 'firebase/firestore';
import { auth, firestore } from '../firebase/config';
import User from './user';

export async function login(email: string, password: string) {
  return await createUserWithEmailAndPassword(auth, email, password);
}

export async function signup(user: User) {
  const userInfo = await createProfile(user.email, user.password, user.names);
  try {
    await setDoc(doc(firestore, 'users', userInfo.uid), {
      email: user.email,
      names: user.names,
    });
  } catch (error) {
    deleteUser(userInfo);
    throw error;
  }
}

export async function getCurrentUser() {
  return await auth.currentUser;
}

async function createProfile(email: string, password: string, names: string) {
  const response = await createUserWithEmailAndPassword(auth, email, password);
  const userInfo = response.user;
  await updateProfile(userInfo, {
    displayName: names,
  });
  return userInfo;
}

File: lib/service/user/user_service.dart
import 'dart:developer';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_auth/firebase_auth.dart' as firebase_auth;
import 'package:theleast/service/auth/auth_service.dart';
import 'package:theleast/service/house/house.dart';
import 'package:theleast/service/user/user.dart';

final db = FirebaseFirestore.instance;

Future<void> createUser(User user, String password) async {
  String? uid = await signup(user.email, password);
  if (uid != null) {
    await db
        .collection("users")
        .doc(uid)
        .set(user.toJson())
        .onError((e, _) => log("Error writing document: $e"));
  }
  log("Creating user Successful");
}

Future<User?> getUser(String userId) async {
  final ref = db.collection("users").doc(userId).withConverter(
        fromFirestore: User.fromFirestore,
        toFirestore: (User house, _) {
          return house.toJson();
        },
      );
  final snapshot = await ref.get();
  return snapshot.data();
}

Future<bool> addFavorite(House house) async {
  final list = await _getFavorites();
  if (!list.contains(house.id)) {
    return await _updateFavorite([...list, house.id]);
  }
  return false;
}

Future<bool> removeFavorite(House house) async {
  final list = List.from(await _getFavorites());
  list.remove(house.id);
  return _updateFavorite(list);
}

Future<bool> _updateFavorite(List<dynamic>? favorites) async {
  String? userId = firebase_auth.FirebaseAuth.instance.currentUser?.uid;

  try {
    await db.collection("users").doc(userId).set(
      {"favoriteHouses": favorites},
      SetOptions(merge: true),
    );
    return true;
  } catch (error, stack) {
    print(stack);
  }
  return false;
}

Future<List> _getFavorites() async {
  String? userId = firebase_auth.FirebaseAuth.instance.currentUser?.uid;
  if (userId == null) {
    return [];
  }

  return (await getUser(userId))?.favoriteHouses ?? [];
}
